src/frontend/src/flows/verifiableCredentials/index.ts

```diff
@@ -0,0 +1,415 @@
+  });
+
+  if (result === "orphan") {
+    await showMessage({
+      message: "No credentials data provided! Wrong URL?",
+    });
+  }
+
+  return await new Promise((_) => {
+    /* halt forever */
+  });
+};
+
+type VerifyCredentials = Parameters<typeof vcProtocol>[0]["verifyCredentials"];
+type VerifyCredentialsArgs = Parameters<VerifyCredentials>[0];
+
+const verifyCredentials = async ({
+  connection,
+  request: {
+    credentialSubject: givenP_RP,
+    issuer: { origin: issuerOrigin, canisterId: issuerCanisterId_ },
+    credentialSpec,
+  },
+  rpOrigin,
+}: { connection: Connection } & VerifyCredentialsArgs) => {
+  // Use the issuer canister ID provided by the RP, or look it up
+  // from the issuer origin if no canister ID was provided.
+  // NOTE: we do _not_ check the origin's canister ID if a canister ID was provided explicitly.
+  let issuerCanisterId = issuerCanisterId_?.toText();
+
+  if (isNullish(issuerCanisterId)) {
+    const lookedUp = await withLoader(() =>
+      lookupCanister({ origin: issuerOrigin })
+    );
+    if (lookedUp === "not_found") {
+      return abortedCredentials({ reason: "no_canister_id" });
+    }
+
+    issuerCanisterId = lookedUp.ok;
+  }
+
+  const vcIssuer = new VcIssuer(issuerCanisterId);
+  // XXX: We don't check that the language matches the user's language. We need
+  // to figure what to do UX-wise first.
+  const consentInfo = await vcIssuer.getConsentMessage({ credentialSpec });
+
+  if (consentInfo === "error") {
+    return abortedCredentials({ reason: "auth_failed_issuer" });
+  }
+
+  // Ask user to confirm the verification of credentials
+  const allowed = await allowCredentials({
+    relyingOrigin: rpOrigin,
+    providerOrigin: issuerOrigin,
+    consentMessage: consentInfo.consent_message,
+    userNumber: undefined,
+  });
+  if (allowed.tag === "canceled") {
+    return "aborted";
+  }
+  allowed.tag satisfies "allowed";
+
+  const userNumber = allowed.userNumber;
+
+  // For the rest of the flow we need to be authenticated, so authenticate
+  // XXX: this simply breaks for PIN identities
+  const authResult = await withLoader(() =>
+    handleLogin({
+      login: () => connection.login(userNumber),
+    })
+  );
+
+  if (authResult.tag !== "ok") {
+    return abortedCredentials({ reason: "auth_failed_ii" });
+  }
+  const authenticatedConnection = authResult.connection;
+
+  // Compute the user's principal on the RP and ensure it matches what the RP sent us
+  const computedP_RP = await withLoader(() =>
+    authenticatedConnection.getPrincipal({
+      origin: rpOrigin,
+    })
+  );
+  if (computedP_RP.compareTo(givenP_RP) !== "eq") {
+    console.error(
+      [
+        "bad principals for user number/origin",
+        userNumber,
+        rpOrigin,
+        computedP_RP.toString(),
+        givenP_RP.toString(),
+      ].join(", ")
+    );
+    return abortedCredentials({ reason: "bad_principal_rp" });
```

Comment:
Instead of cancelling and letting the RP know, shouldn't there be a retry option on II?